src/backend/distributed/connection/remote_commands.c

```diff
@@ -531,3 +555,257 @@ GetRemoteCommandResult(MultiConnection *connection, bool raiseInterrupts)
+		int socket = PQsocket(connection->pgConn);
+		int eventMask = WL_SOCKET_READABLE;
+		bool connectionFailed = false;
+
+		if (PQstatus(connection->pgConn) == CONNECTION_OK)
+		{
+			int sendStatus = PQflush(connection->pgConn);
+			if (sendStatus == -1)
+			{
+				connectionFailed = true;
+			}
+			else if (sendStatus == 1)
+			{
+				/* more data to flush */
+				eventMask |= WL_SOCKET_WRITEABLE;
+			}
+		}
+		else
+		{
+			connectionFailed = true;
+		}
+
+		if (!connectionFailed)
+		{
+			connectionReady[connectionIndex] = false;
+			AddWaitEventToSet(waitEventSet, eventMask, socket, NULL, (void *) connection);
+		}
+		else
+		{
+			connectionReady[connectionIndex] = true;
+			foundReadyConnections = true;
+		}
+	}
+
+	AddWaitEventToSet(waitEventSet, WL_POSTMASTER_DEATH, PGINVALID_SOCKET, NULL, NULL);
+	AddWaitEventToSet(waitEventSet, WL_LATCH_SET, PGINVALID_SOCKET, MyLatch, NULL);
+
+	while (!foundReadyConnections)
+	{
+		int eventIndex = 0;
+		int eventCount = 0;
+		long timeout = 1000;
+
+		/* wait for I/O events for up to 1 second */
+#if (PG_VERSION_NUM >= 100000)
+		eventCount = WaitEventSetWait(waitEventSet, timeout, waiter->events,
+									  pendingConnectionCount, WAIT_EVENT_CLIENT_READ);
+#else
+		eventCount = WaitEventSetWait(waitEventSet, timeout, waiter->events,
+									  pendingConnectionCount);
+#endif
+		if (eventCount == 0)
+		{
+			/* timeout, try again */
+			continue;
+		}
+
+		for (; eventIndex < eventCount; eventIndex++)
+		{
+			WaitEvent *event = &waiter->events[eventIndex];
+			MultiConnection *connection = NULL;
+			bool connectionIsReady = false;
+
+			if (event->events & WL_POSTMASTER_DEATH)
+			{
+				ereport(ERROR, (errmsg("postmaster was shut down, exiting")));
+			}
+
+			if (event->events & WL_LATCH_SET)
+			{
+				ResetLatch(MyLatch);
+
+				/* if allowed, raise errors */
+				if (raiseInterrupts)
+				{
+					CHECK_FOR_INTERRUPTS();
+				}
+
+				if (InterruptHoldoffCount > 0 && (QueryCancelPending || ProcDiePending))
+				{
+					cancelled = true;
+					break;
+				}
+
+				continue;
+			}
+
+			connection = (MultiConnection *) event->user_data;
+			connectionIndex = event->pos + pendingConnectionsStartIndex;
+
+			if (event->events & WL_SOCKET_READABLE)
+			{
+				int receiveStatus = PQconsumeInput(connection->pgConn);
+				if (receiveStatus == 0)
+				{
+					/* receive failed, done with this connection */
+					connectionIsReady = true;
+				}
+				else if (!PQisBusy(connection->pgConn))
+				{
+					/* result was received */
+					connectionIsReady = true;
+				}
+			}
+			if (event->events & WL_SOCKET_WRITEABLE)
+			{
+				int sendStatus = PQflush(connection->pgConn);
+				if (sendStatus == -1)
+				{
+					/* send failed, done with this connection */
+					connectionIsReady = true;
+				}
+				else if (sendStatus == 0)
+				{
+					/* done writing, only wait for read events */
+					ModifyWaitEvent(waitEventSet, connectionIndex, WL_SOCKET_READABLE,
+									NULL);
+				}
+			}
```

Comment:
I don't see why that is? It seems perfectly reasonable to want to wait till a command is sent, but hasn't returned yet.  I think this basically is a sign that the abstraction what to wait for ain't quite right.